### PR TITLE
fix(ops): stop the drift gate aborting on its first finding (#2723)

### LINKED ISSUE #2723: fix(ops): set -e plus ((DRIFT_ERRORS++)) aborts what-matters-now at the first drift, so the required drift gate's entire diagnostic block is unreachable
Found during the bounded review of #2688 and verified independently. **Pre-existing and correctly out of scope for #2688** (fails the `introduced_here` condition in `delivery.json` `blocker_predicate`), but it disables the diagnostic output of a **required** CI context, so it needs its own owner.

## The defect

`ops/scripts/what-matters-now.sh` sets `set -euo pipefail` (`:14`) and then counts drift with post-increment:

```sh
((DRIFT_ERRORS++))     # :30, :53, :75, :80, :84, :115
```

`((expr))` returns exit status **1 when the expression evaluates to 0**. Post-increment evaluates to the value *before* the increment — so the very **first** drift, `0 → 1`, yields status 1 and `set -e` **aborts the script on the spot**.

Reproduced minimally:

```
$ cat t.sh
set -euo pipefail
N=0
echo "before: N=$N"
((N++))
echo "AFTER FIRST INCREMENT: N=$N"

$ bash t.sh
before: N=0
script exit=1          # the second echo never runs
```

## What this actually costs

The script aborts at the first `check_file` miss, so everything downstream is **unreachable whenever drift exists**:

```sh
# :204-212 — dead code in the drift case
if [[ ${DRIFT_ERRORS} -gt 0 ]]; then
  echo "DRIFT DETECTED: ${DRIFT_ERRORS} problem(s)"
  [[ "${TRUTH_OK}" == "false" ]] && echo "  Missing canonical truth files"
  [[ "${SYMLINKS_OK}" == "false" ]] && printf "${SYMLINK_WARNINGS}"
  [[ -n "${STALE_HITS}"      ]] && printf "Stale paths:\n${STALE_HITS}"
  exit 1
fi
```

Consequences:

1. **`--drift` never explains itself.** `Agent Tooling Drift Check` is one of the 11 **required** contexts. The moment it has something real to say, an operator gets a bare non-zero exit and a `MISSING: <file>` line on **stderr** — never the `DRIFT DETECTED: N problem(s)` summary, never the missing-truth-files line, never the symlink warnings, never the stale-path list. The gate fires correctly and reports almost nothing.
2. **`--json` emits zero bytes on any drift**, independently of the heredoc defect #2688 fixes. Verified by copying `ops/` to a temp tree and deleting `ops/state/truth/agents.json`: `--json rc=1 stdout_len=0`, `--drift rc=1 stdout_len=0`.
3. **`exit ${DRIFT_ERRORS}`** (`:202`) can only ever run with the value `0`, so the documented "exit status equals the drift count" contract is unobservable above zero.

This is a *silent* failure rather than a false success, but it corrupts evidence the same way: the check that exists to explain drift is the one that goes quiet when drift appears.

## Fix direction

Use a form whose status does not depend on the value — `DRIFT_ERRORS=$(( DRIFT_ERRORS + 1 ))`, or `((DRIFT_ERRORS++)) || true`, or pre-increment `((++DRIFT_ERRORS))` (which returns the *new* value and so only misfires at −1). The first is clearest and has no edge case at all.

## Acceptance criteria

- [ ] All six increment sites use a form that cannot abort under `set -e`.
- [ ] A test injects real drift (e.g. a missing canonical truth file in a temp tree) and asserts the `--drift` output **contains** `DRIFT DETECTED` and the specific diagnostic lines — not merely that the exit status is non-zero. An exit-status-only assertion passes today and would not discriminate.
- [ ] A test asserts `--json` still emits parseable JSON when drift is present, and that `drift_errors` is the true count.
- [ ] A test asserts the exit status equals the reported `drift_errors` for a **non-zero** count. (#2688's test names this contract at `test_what_matters_now_json.py:76` but can currently only observe `0 == 0` — see the note below.)

## Relationship to #2688

#2688 fixes a different defect on the same path — the unquoted Python heredoc — and its test suite is the right place for the coverage above. Its existing exit-status check cannot reach the interesting case until *this* issue is fixed; that limitation was recorded in review as FOLLOW_UP, not as a blocker, and #2688 should not be widened to absorb this.

Related pre-existing interpolation debt in the same file: **#2722**.

Refs #2688 #2638 #2722.

### PR DESCRIPTION
**Pull request by fahertym**
Closes #2723.

```
ICN DELIVERY LIFECYCLE
State:                FROZEN  (11/11 required contexts green on the exact head;
                      awaiting explicit human merge authorization)
Lane:                 FAST -- required-gate observability
Acceptance contract:  when drift exists the script must continue through every check, count
                      each finding, reach the final report, print bounded actionable
                      diagnostics, and exit non-zero. When none exists it must inspect every
                      check, report clean, and exit zero.
Review generation:    1 bounded FULL review -> 1 BLOCKER, 2 FOLLOW_UP, 0 QUESTION,
                      6 NOT_A_FINDING. The blocker (a newline regression introduced by this
                      PR's own printf fix) and the coverage gap that hid it are both fixed in
                      2cfcc29c4. Generation CLOSED; review is now DELTA only.
Freeze head:          2cfcc29c418afc5e4d796db35496a0c2bc64f979
Base:                 main dd21b2d008963dad91cc9b1c8c485133680115a3 -- merge-base equals
                      origin/main (strict:true satisfied)
Known blockers:       none open
Unresolved threads:   0 (3 Copilot threads: two valid and fixed, one declined with evidence)
Non-required red:     Security Audit only (icn#2579) -- red on main itself, not caused here
Follow-up ledger:     (1) icn#2722 -- ${SPRINT_FILE} still spliced into python3 source in this
                          same file; deliberately untouched, separate owner
                      (2) check_file() has no callers, so one of the six corrected increments
                          is dead code. Pre-existing; fails `introduced_here`. Deleting a
                          helper is a separate decision from making a counter safe
```

## The defect

`Agent Tooling Drift Check` is one of the **11 required contexts**, and `ops/scripts/what-matters-now.sh --drift` is what it runs.

Under `set -euo pipefail` the count was kept with `((DRIFT_ERRORS++))`. An arithmetic **command** whose value is zero exits 1, and post-increment evaluates to the value *before* the increment — so the very first drift, `0 → 1`, aborted the script.

Everything downstream became unreachable exactly when it mattered: the `DRIFT DETECTED: N problem(s)` report, the per-category detail lines, and `exit ${DRIFT_ERRORS}` — which could only ever run with `0`.

## Reproduced before changing anything

A git fixture carrying the real `ops/` tree, with one canonical truth file removed, run against the **pre-fix** script:

```
exit 1
stdout: (empty)
stderr: (empty)
```

**A required check went red and said nothing at all.** That is worse than a failing check — it is a control that stops before producing the evidence it exists to produce. The truth-file loop counts *silently* (the `MISSING:` message lives in `check_file`, which that loop does not use), so there was not even a partial clue.

## The fix

All six increments become `DRIFT_ERRORS=$(( DRIFT_ERRORS + 1 ))`, whose exit status does not depend on the value. The declaration carries a note so the form is not "simplified" back later.

Two repairs the same acceptance contract requires:

- the canonical-truth loop now **records which files are missing**, so the report names them instead of saying only "Missing canonical truth files";
- the test is wired into the workflow **beside the gate it guards** — a mode with no runner is how the original defect survived, and #2688 had just established that pattern.

## Behaviour now

```
clean tree   ->  "OK: no drift detected"                                    exit 0
one drift    ->  "DRIFT DETECTED: 1 problem(s)" + the file named            exit 1
two drifts   ->  "DRIFT DETECTED: 2 problem(s)" + both named                exit 1
```

**Aggregation is the load-bearing case.** It was impossible before, because the first finding ended the run.

## Proof the tests discriminate

`scripts/tests/test_what_matters_now_drift.py` — 18 checks, all passing. With only the six increments reverted, **7 fail**, including both behavioural assertions reporting `output=''`.

The controls assert the shell semantics rather than assuming them: `((N++))` with `N=0` under `set -e` is *executed* and shown to abort, and the assignment form is shown not to. The matcher is also proved to recognise the pre-fix form and to **not** flag the safe form, so it cannot pass by being blind.

Per the issue's C3 requirement, the structural check rejects `((VAR++))`, `((VAR--))`, `((++VAR))` **and** `((--VAR))` anywhere in the script — a different unsafe counter cannot be introduced later on the same executed path. I swept `ops/scripts/` for the same hazard: **`what-matters-now.sh` is the only `set -e` script that carried it.**

## Non-goals

- **Does not weaken the required context.** Drift still exits non-zero; it now explains itself.
- **#2722 is untouched** — `${SPRINT_FILE}` is still spliced into `python3` source in this same file. Different defect, separate owner, deliberately not folded in.
- No change to what counts as drift, only to whether the script survives long enough to report it.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: scripts/tests/test_what_matters_now_drift.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+Measured on the pre-fix script: one missing canonical truth file produced exit 1
+with **completely empty stdout and stderr** -- the truth-file loop counts without
+printing, so a required check went red and said nothing at all.
+
+The controls below fail against the pre-fix form. A gate that cannot fail is not
+a gate.
+"""
+
+import os
```
[the remark this answers is not in the document]
Comment (fahertym) [reply scripts/tests/test_what_matters_now_drift.py]: Correct — fixed. `os` was left over from an earlier draft that built the fixture by hand before switching to `shutil.copytree`.

File: ops/scripts/what-matters-now.sh [hunk 2]
```diff
@@ -258,7 +268,10 @@ fi
 if [[ "${MODE}" == "--drift" ]]; then
   if [[ ${DRIFT_ERRORS} -gt 0 ]]; then
     echo "DRIFT DETECTED: ${DRIFT_ERRORS} problem(s)"
-    [[ "${TRUTH_OK}" == "false" ]] && echo "  Missing canonical truth files"
+    if [[ "${TRUTH_OK}" == "false" ]]; then
+      echo "  Missing canonical truth files:"
+      printf '%s' "${MISSING_TRUTH_FILES}"
+    fi
     [[ "${SYMLINKS_OK}" == "false" ]] && printf "${SYMLINK_WARNINGS}"
     [[ -n "${STALE_HITS}" ]] && printf "Stale paths:\n${STALE_HITS}"
```
[the remark this answers is not in the document]
Comment (fahertym) [reply ops/scripts/what-matters-now.sh]: Correct, fixed, and it matters more than it looks — thank you.

Those two lines were **unreachable dead code until this PR**. The abort meant nothing after the first drift ever ran, so a format-string bug in the drift report could not fire. By making the report reachable I made this bug reachable too, which puts it squarely in scope rather than adjacent to it.

Demonstrated:

```
$ S=$(printf "100%%%% done\n50%%s here"); printf "$S"
100% done
50 here          <- `%%` collapsed AND `%s` silently ate content
$ printf "%s" "$S"
100%% done
50%s here        <- verbatim
```

Content *disappearing* is the failure mode that matters here: `STALE_HITS` carries grep output over real paths, and `SYMLINK_WARNINGS` carries link targets. A diagnostic that exists to tell an operator what is wrong must not drop part of the answer because a path contained a `%`.

Fixed at all four sites, not just the two in the drift report — `printf "  ✗ Problems:\\n${SYMLINK_WARNINGS}"` and `printf "  ✗ Stale path hits:\\n${STALE_HITS}"` in human mode had the same shape and are reachable today. All now use a literal format with `%s`.

File: scripts/tests/test_what_matters_now_drift.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+print("--- no counter in this script can abort under `set -e` ---")
+script = (REPO_ROOT / SCRIPT_REL).read_text()
+
+# `((VAR++))` / `((VAR--))` as a *command* is the hazard: its exit status is 1
+# whenever the expression evaluates to 0. Assignment forms are safe.
+UNSAFE = re.compile(r"^\s*\(\(\s*[A-Za-z_][A-Za-z0-9_]*\s*(\+\+|--)\s*\)\)", re.M)
+UNSAFE_PREFIX = re.compile(r"^\s*\(\(\s*(\+\+|--)\s*[A-Za-z_][A-Za-z0-9_]*\s*\)\)", re.M)
+
+check("the script sets -e (otherwise this whole class is moot)",
+      re.search(r"^set -[a-z]*e", script, re.M) is not None)
```
[the remark this answers is not in the document]
Comment (fahertym) [reply scripts/tests/test_what_matters_now_drift.py:151]: Checked, and this one is not right — no change made.

`^set -[a-z]*e` does match `set -euo pipefail`. `[a-z]*` can match the empty string, after which `e` matches the first character of `euo`:

```
$ python3 -c "import re; print(bool(re.search(r'^set -[a-z]*e', 'set -euo pipefail', re.M)))"
True
```

The check also passes in the live run — `ok   the script sets -e (otherwise this whole class is moot)` — against the real script, which is the only string it is ever applied to.

Leaving as is rather than loosening a regex that already matches, since a broader pattern would weaken the guard for no gain.

**Comment by fahertym** [comment]
## Bounded FULL review — findings dispositioned

```
Kind:      FULL
Findings:  1 BLOCKER · 2 FOLLOW_UP · 0 QUESTION · 6 NOT_A_FINDING
```

### BLOCKER — valid, and self-inflicted. Fixed in `2cfcc29c4`.

The review caught a regression **I introduced in `134c582dc`** while fixing the printf format-string hazard. Switching to `printf '%s' "${VAR}"` removed the format-string bug *and* the escape processing the buggy form was accidentally providing. The accumulators append a two-character `\n`, interpreted only because the variable *was* the format. With a literal `%s`, nothing interprets it:

```
NOT SYMLINK: ...\n  MISSING: ...\n  MISSING: ...\nStale paths:
```

Every finding except the truth-file one collapsed onto one run-on line — in the report this PR exists to make reachable. The predicate is met on every condition, and the reviewer was right to call it a blocker rather than a nit.

Fixed by appending a real newline via `$'\n'` in all five accumulators, matching what `MISSING_TRUTH_FILES` already did correctly here. **`%b` deliberately not used** — it would have handled the escapes and reintroduced backslash processing over grep output, which is the trap the reviewer explicitly flagged.

### FOLLOW_UP 1 — coverage gap. **Fixed here rather than deferred**, because it is the reason the blocker landed.

All 18 checks passed at `134c582dc` *with the rendering broken*. Nothing exercised the symlink or stale-path channels — a gate blind to two of the three things it guards. Deferring that would have left the suite able to miss the same class again.

The fixture now triggers all three phases (a plain directory where a skill symlink belongs, two absent skills, a stale pattern in a scanned file) and asserts each finding lands on its own line with no literal backslash-n anywhere in the output.

Mutation evidence for the new coverage: reverting the accumulators fails **"no literal backslash-n leaks into the report"** and **"symlink findings are on their own lines"**.

Two controls now pin the printf semantics by *executing* them — `printf "$VAR"` shown to eat content (`a%sb` → `ab`), `printf '%s' "$VAR"` shown to preserve it — plus a structural check that no variable is ever passed as a format anywhere in the script.

I also removed a placeholder check that passed unconditionally. A check that cannot fail is precisely what this PR is about; leaving one in would have been the same mistake in miniature.

### FOLLOW_UP 2 — `check_file()` is dead code. Not fixed; correctly out of scope.

It has no callers, so one of the six increments I corrected is unreachable. Pre-existing, fails `introduced_here`. Recorded here rather than removed, since deleting a helper is a separate decision from making a counter safe.

### NOT_A_FINDING ×6 — confirmed

Subshell scoping (all six increments run in the main shell; the `while read <<<` is a here-string, not a pipe — a three-phase drift aggregated to `5 problem(s)`) · `[[ ]] && printf` cannot exit under `set -e` in an AND-list · `MISSING_TRUTH_FILES` is safe under `set -u` · exit-status wrap is unreachable (max count 21) and pre-existing · the workflow step has no `continue-on-error`, so it fails the required context · test hygiene and drift conditions unchanged.

**Head moved to `2cfcc29c4`; the earlier green does not cover it.** Required CI is re-running, and only that result will be used to freeze.

**Comment by fahertym** [comment]
# Merge-decision packet — #2728 (fixes #2723)

**Status: FROZEN. Ready for a human merge decision. Not merged — no authorization present in this session.**

| | |
|---|---|
| **Freeze head** | **`2cfcc29c418afc5e4d796db35496a0c2bc64f979`** |
| Base | `main` = **`dd21b2d008963dad91cc9b1c8c485133680115a3`** — merge-base **equals** `origin/main`, `strict: true` satisfied |
| Required CI | **11/11 SUCCESS** on this exact head |
| Threads | **0 unresolved** (3 Copilot threads: two valid and fixed, one declined with evidence) |
| Diff | 3 files — shell + Python + workflow. No Rust. |

Only non-required red is `Security Audit` (#2579), red on `main` itself.

## What it fixes

`Agent Tooling Drift Check` is a **required** context. Under `set -euo pipefail`, `((DRIFT_ERRORS++))` exits 1 when the value is 0, and post-increment yields the pre-increment value — so the **first** drift aborted the script.

Measured pre-fix with one canonical truth file removed:

```
exit 1, stdout and stderr both completely empty
```

A required check going red and saying **nothing at all** — a control that stops before producing the evidence it exists to produce.

**After:** clean → exit 0 · one drift → `DRIFT DETECTED: 1 problem(s)` + the file named · two drifts → **aggregates to 2**, both named. Aggregation was impossible before.

## Review, including a blocker this PR created

**1 BLOCKER, 2 FOLLOW_UP, 6 NOT_A_FINDING.**

The blocker was **self-inflicted**: fixing the printf format-string hazard also removed the escape processing the buggy form was accidentally providing, so the accumulators' two-character `\n` stopped rendering and every finding except the truth-file one collapsed onto one run-on line — in the report this PR exists to make reachable. Fixed with real newlines via `$'\n'`. **`%b` deliberately not used**; it would have handled the escapes while reintroducing backslash processing over grep output.

The coverage gap that hid it was fixed here rather than deferred, because it is *why* the blocker landed: all 18 checks passed with the rendering broken, since nothing exercised the symlink or stale-path channels. The fixture now drives all three phases.

I also removed a placeholder check that passed unconditionally — a check that cannot fail is precisely what this PR is about.

**One review comment was declined with evidence**: the claim that `^set -[a-z]*e` cannot match `set -euo pipefail`. It can (`[a-z]*` matches empty, `e` matches the first character of `euo`), verified empirically and passing against the real script every run. Loosening a working guard would weaken it for nothing.

## Proof the tests discriminate

22 checks. Mutation on the counter fails 7; mutation on the accumulators fails 2 more. Controls **execute** the semantics rather than asserting them — `((N++))` with `N=0` shown to abort under `set -e`, `printf "$VAR"` shown to eat content (`a%sb` → `ab`), and the safe forms shown not to. The matcher is proved to recognise the unsafe form *and* not flag the safe one, so it cannot pass by being blind.

Per the issue's C3 requirement it rejects `((VAR++))`, `((VAR--))`, `((++VAR))` and `((--VAR))` anywhere in the script. `what-matters-now.sh` was the only `set -e` script under `ops/scripts/` carrying the hazard.

## Non-claims

- **Does not weaken the required context.** Drift still exits non-zero; it now explains itself.
- **#2722 untouched** — `${SPRINT_FILE}` is still spliced into `python3` source in this same file.
- `check_file()` is dead code, so one corrected increment is unreachable. Pre-existing, recorded not removed.
- No change to what counts as drift.

## Decision required

**Merge authorization for #2728 at exactly `2cfcc29c418afc5e4d796db35496a0c2bc64f979`.** Squash, no `--admin`. Closes #2723.
